staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go

```diff
@@ -250,21 +322,74 @@ func (s *openAPIAggregator) loadApiServiceSpec(handler http.Handler, apiService
 		return nil
 	}
 
-	openApiSpec, err := s.downloadOpenAPISpec(handler)
+	openApiSpec, etag, _, err := s.downloadOpenAPISpec(handler, "")
 	if err != nil {
 		return err
 	}
+	if openApiSpec == nil {
+		// API service does not provide an OpenAPI spec, ignore it.
+		return nil
+	}
 	aggregator.FilterSpecByPaths(openApiSpec, []string{"/apis/" + apiService.Spec.Group + "/"})
 
+	_, existingService := s.openAPISpecs[apiService.Name]
 	s.openAPISpecs[apiService.Name] = &openAPISpecInfo{
 		apiService: *apiService,
 		spec:       openApiSpec,
+		handler:    handler,
+		etag:       etag,
 	}
 
 	err = s.updateOpenAPISpec()
 	if err != nil {
 		delete(s.openAPISpecs, apiService.Name)
 		return err
 	}
+	if !existingService {
+		s.openAPIAggregationController.enqueue(apiService.Name)
+	}
+
 	return nil
 }
+
+func (s *openAPIAggregator) removeApiServiceSpec(apiServiceName string) error {
+	if _, existingService := s.openAPISpecs[apiServiceName]; existingService {
+		oldSpecs := s.openAPISpecs
+		delete(s.openAPISpecs, apiServiceName)
+		err := s.updateOpenAPISpec()
+		if err != nil {
+			s.openAPISpecs = oldSpecs
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *openAPIAggregator) UpdateApiServiceSpec(apiServiceName string) (shouldBeRateLimited, deleted bool, err error) {
+	specInfo, existingService := s.openAPISpecs[apiServiceName]
+	if !existingService {
+		return false, true, nil
+	}
+	// If this is a local spec with no handler, that means it is a static spec.
+	if specInfo.handler == nil {
+		return false, false, nil
+	}
+	spec, etag, httpStatus, err := s.downloadOpenAPISpec(specInfo.handler, specInfo.etag)
+	if err != nil {
+		return true, false, err
+	}
+	if httpStatus == http.StatusNotModified {
+		return false, false, nil
+	}
+	if httpStatus == http.StatusNotFound || spec == nil {
+		return true, false, nil
+	}
+	oldSpec := specInfo.spec
+	specInfo.spec = spec
+	if err := s.updateOpenAPISpec(); err != nil {
+		specInfo.spec = oldSpec
```

Comment:
I tried to keep the logic that if a function fails, it won't change the state of anything. Here I keep old spec and restore it if we fail to update new one.